src/main/java/com/example/petbuddybackend/utils/specification/CaretakerSpecificationUtils.java

```diff
@@ -0,0 +1,91 @@
+    public static final String CITY = "city";
+    public static final String ACCOUNT_DATA = "accountData";
+    public static final String NAME = "name";
+    public static final String SURNAME = "surname";
+    public static final String ANIMAL_TYPES = "animalsTakenCareOf";
+    public static final String EMAIL = "email";
+
+
+    private CaretakerSpecificationUtils() {
+    }
+
+    public static Specification<Caretaker> toSpecification(CaretakerSearchCriteria filters) {
+        var spec = Specification.<Caretaker>where(
+                (root, query, criteriaBuilder) -> criteriaBuilder.conjunction()
+        );
+
+        if (StringUtils.hasText(filters.personalDataLike())) {
+            spec = spec.and(personalDataLike(filters.personalDataLike()));
+        }
+
+        if (StringUtils.hasText(filters.cityLike())) {
+            spec = spec.and(cityLike(filters.cityLike()));
+        }
+
+        if (filters.voivodeship() != null) {
+            spec = spec.and(voivodeshipEquals(filters.voivodeship()));
+        }
+
+        if (!CollectionUtils.isEmpty(filters.animalTypes())) {
+            spec = spec.and(animalTypesIn(filters.animalTypes()));
+        }
+
+        return spec;
+    }
+
+    private static Specification<Caretaker> personalDataLike(String personalDataLike) {
+        return (root, query, criteriaBuilder) -> {
+            String[] keywords = personalDataLike.toLowerCase().split("\\s+");
+            List<Predicate> predicates = new ArrayList<>(keywords.length);
+
+            for (String keyword : keywords) {
+                String likePattern = "%" + keyword + "%";
+                predicates.add(
+                        criteriaBuilder.or(
+                                criteriaBuilder.like(criteriaBuilder.lower(root.get(ACCOUNT_DATA).get(NAME)), likePattern),
+                                criteriaBuilder.like(criteriaBuilder.lower(root.get(ACCOUNT_DATA).get(SURNAME)), likePattern),
+                                criteriaBuilder.like(criteriaBuilder.lower(root.get(ACCOUNT_DATA).get(EMAIL)), likePattern)
+                        )
+                );
+            }
+
+            return criteriaBuilder.and(predicates.toArray(new Predicate[0]));
+        };
+    }
+
+    private static Specification<Caretaker> cityLike(String cityLike) {
+        return (root, query, criteriaBuilder) -> {
+            String likePattern = "%" + cityLike.toLowerCase() + "%";
+            return criteriaBuilder.like(criteriaBuilder.lower(root.get(ADDRESS).get(CITY)), likePattern);
+        };
+    }
+
+    private static Specification<Caretaker> voivodeshipEquals(Voivodeship voivodeship) {
+        return (root, query, criteriaBuilder) ->
+                criteriaBuilder.equal(root.get(ADDRESS).get(VOIVODESHIP), voivodeship);
+    }
+
+    private static Specification<Caretaker> animalTypesIn(Set<AnimalType> animalTypes) {
+        return (root, query, criteriaBuilder) ->
+                root.join(ANIMAL_TYPES).in(animalTypes);
```

Comment:
i jesli bysmy zrobili osobna tabele Animals to musialaby powstac tabela laczaca Caretaker i Animal, wtedy tam kluczem moze byc email, nazwa_zwierzecia(pies, kot) i raczej tutaj nie bedzie problemu znaleźć też